docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,2979 @@ p4grammar {
+
+- $\ruleNameT$ indicates that the rule is describing a rule for a type
+- $\ruleNameE$ and $\ruleNameAE$ indicate that the rules are for an expression and an auxiliary judgment for an expression
+- $\ruleNameS$ indicates that the rule is describing a rule for a statement
+- $\ruleNameD$ indicates that the rule is describing a rule for a declaration
+
+## rest
+
+``TODO: add followings.``
+
+- $\overline {\overline \anyTyp}$
+- $\overline {X:\typ}$ zips $\overline X$ and $\typs$.
+- $\concatList a b$
+
+Metavariables are place holders for values of a type. For example, $\bool$ stands for expressions $true$ or $false$. Keep in mind that $true$ and $false$ both are expressions that have the type $\boolTyp$. As another example, in most programming references $foo$, $bar$, and $baz$ are used as metavariables, often for functions. 
+
+- $\overline x$ denotes a list of $x$.
+- For simplicity, we omit mapping a function/judgment on a list of elements. Instead, we just apply the function/judgment to a list. So every time you see a function applied to a list, read it as mapping the function over all the elements of the list. 
+- An underscore (that is, $\_$) is used when we do not need to use some specific information returned by some helper in judgments. 
+- The notation $\maybe \anyTyp$ denotes an optional type, that is, the type of $\anyTyp$ has been extended with bottom (maybe type in Haskell, or option type in OCaml). So if $\anyTyp$ denotes an integer, $\maybe \anyTyp$ could either be an integer number or bottom. 
+
+The following describes the metavariables used in this document:
+
+~ Center
+| **metavariable**   |   **type represented** |
+|:----------------:|:------------------|
+| $\bool$ | booleans |
+| $\str$ | strings |
+| $\int$ | arbitrary length integer |
+| $\width$ | width of a data structure |
+| $\size$ | size of a data structure |
+| $\bitWidth \bit \width$ | fixed length integer |
+| $\intWidth \int \width$ | signed integer |
+| $\name$ | strings representing any kinds of names |
+| $\typ$ | types |
+| $\dir$ | directions |
+| $\typVar$ | type variables |
+| $\wildcardParam$ | wild card type parameters |
+| $\field$ | field names |
+| $\env$ | environments |
+| $\emp$ | empty list of elements |
+| $\anyTyp$ | a value of any type |
+| $\constraint$, $\optConstraint$, $\nonoptConstraint$ | constraints |
+~
 
-1. It type checks P4 programs.
-2. It conducts type inference.
-3. It does a pass from the surface syntax to the first IR. 
 
+## Signatures
+``TODO``
+function signature. judgment signature. environment signature. 
+
+## contexts and environments
+
+# Syntax { #sec-syntax }
+
+In the following we present P4 syntax. Note that we omitted annotations and meta-information since they are not required for type checking.
+
+~ Begin P4Syntax
+**Data types used in P4 programs:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\baseTyp$ | $::=$ | $\boolTyp$ | $\qquad \text{booleans}$ |
```

Comment:
What's the difference between $\basetype$ and $\typ$? It's not totally clear to me from the descriptions, "Data types used in P4 programs" vs. "Types of P4 expressions". Seems like the former should be a superset of the latter, but it is not.

Reply:
`\typ` is the type of expression that is produced by the type system whereas `\basetyp` is the type that appears in a P4 expression.

Connecting it to Petr4 implementation, `\typ` is the type used at the IR level and `\basetyp` is the type used at the surface level.

Reply:
break up `\typ` into different syntactic categories. 
X = type variables
x = names assigned to types declared in a program
1) surface level
2) declared types that cannot be referred 

drop W. add wildcard to parameter. drop k. 
check function type, you need parameter directions and types. 
think about including extern methods within the type. 
distinguish between records/...

Reply:
- [ ] in syntax, I noted that a type variable can also be a wildcard. Does the syntax need to represent this? we can have:
`X ::= x | _ `